src/vec/select.php

```diff
@@ -231,6 +231,54 @@ function keys<Tk, Tv>(
   return $result;
 }
 
+/**
+ * Returns the last element of a given vec and removes it from the vec.
+ * Throws if the given vec is empty.
+ * Previously known as `array_pop` in PHP.
+ *
+ * @see `Vec\pop_front()` to remove the first element instead.
+ *
+ * Time complexity: O(either 1 or n) 1 if no other references are being held.
+ *   n if the vec needs to be copied.
+ * Space complexity: O(either 1 or n) 1 if no other references are being held.
+ *   n if the vec needs to be copied.
+ */
+<<__NonRx('Mutates the argument it is given')>>
+function pop_back<T>(inout vec<T> $vec): T {
+  invariant(
+    !C\is_empty($vec),
+    '%s: Expected at least one element.',
+    __FUNCTION__,
+  );
+  /* HH_FIXME[2049] __PHPStdLib */
+  /* HH_FIXME[4107] __PHPStdLib */
+  return \array_pop(inout $vec);
+}
+
+/**
+ * Returns the first element of a given vec and removes it from the vec.
+ * Throws if the given vec is empty.
+ * Previously known as `array_shift` in PHP.
+ *
+ * @see `Vec\pop_back()` to remove the last element instead.
+ *
+ * Time complexity: O(either 1 or n) 1 if no other references are being held.
+ *   n if the vec needs to be copied.
+ * Space complexity: O(either 1 or n) 1 if no other references are being held.
+ *   n if the vec needs to be copied.
+ */
```

Comment:
Is this correct? I though it would be always O(n) because it has to move every element to the left by 1.

Reply:
I think I might have taken an incorrect mental shortcut when writing the comment. `array_shift` could be constant time, if a machanism was used where the first element is decreffed and the storage repointed to the second element. However I have no idea how this got in my head.
Even worse, you can't do this on `array_unshift` at all, since you couldn't safely steal space in front to repoint to.